---
title: "R Code"
author: "A BUTT"
date: "07/01/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---
title: "RNAseq_results_DEA"
author: "A BUTT"
date: "07/01/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set working directory 
```{r}
dir="\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq"
files=list.files(dir,full=TRUE)
list.files(dir)
```
```{r}
getwd()
```

Load library
```{r}
library(rlang)
library(dplyr)
library(ggplot2)
library(DESeq2)
library(reshape)
library(reshape2)
library(janitor)
library(edgeR)
library(limma)
library(EnhancedVolcano) 
library(tidyverse)
library(caret)
library(scales) 
library(viridis)
library(ggdendro)
library(gridExtra)
library(gtable)
library(grid)
library(ggplot2)
library(fgsea)
library(tibble)
library(matrixStats)
library(apeglm)
library(VennDiagram)
library(sqldf)
library(rstatix)
library(ggpubr)
library(RColorBrewer)
library(SummarizedExperiment)
library(pheatmap)
library(PoiClaClu)
library(ashr)

```

CIBERSORTx results 
- creating boxplots for different monocyte/macrophage populations

M2 macrophage population
```{r}
cibersort <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\CIBERSORTx_Job6_Results_BOXPLOTS.txt")
head(cibersort)

cibersort2 <- cibersort
class(cibersort$Timepoint)
cibersort2$Timepoint <- as.factor(cibersort2$Timepoint)
class(cibersort2$Timepoint)

levels(cibersort2$Timepoint)

#M2 population

#Wilcoxon matched-pairs signed rank test
cibersort2 %>%
  group_by(Timepoint) %>%
  get_summary_stats(Macrophages.M2, type = "median_iqr")

bxp <- ggboxplot(
  cibersort2, x = "Timepoint", y = "Macrophages.M2", color = "Timepoint", palette = c("#00AFBB", "#E7B800", "#FC4E07"), FILL = "white", ylab = "M2 population", xlab = "Timepoint", add = "jitter", shape = "Timepoint"
)
bxp

stat.test <- cibersort2 %>%
  pairwise_wilcox_test(Macrophages.M2 ~ Timepoint) %>%
  add_significance()
stat.test

stat.test <- stat.test %>% add_xy_position(x = "Timepoint")
bxp +
  stat_pvalue_manual(stat.test, tip.length = 0) +
  labs(subtitle = get_test_label(stat.test, detailed = TRUE))

```

M0 macrophage population
```{r}
#M0 population

#Wilcoxon matched-pairs signed rank test
cibersort2 %>%
  group_by(Timepoint) %>%
  get_summary_stats(Macrophages.M0, type = "median_iqr")

bxp <- ggboxplot(
  cibersort2, x = "Timepoint", y = "Macrophages.M0", color = "Timepoint", palette = c("#00AFBB", "#E7B800", "#FC4E07"), FILL = "white", ylab = "M0 population", xlab = "Timepoint", add = "jitter", shape = "Timepoint"
)
bxp

stat.test <- cibersort2 %>%
  pairwise_wilcox_test(Macrophages.M0 ~ Timepoint) %>%
  add_significance()
stat.test

stat.test <- stat.test %>% add_xy_position(x = "Timepoint")
bxp +
  stat_pvalue_manual(stat.test, tip.length = 0) +
  labs(subtitle = get_test_label(stat.test, detailed = TRUE)) 
```

Monocyte population
```{r}
#Monocyte population

#Wilcoxon matched-pairs signed rank test
cibersort2 %>%
  group_by(Timepoint) %>%
  get_summary_stats(Monocytes, type = "median_iqr")

bxp <- ggboxplot(
  cibersort2, x = "Timepoint", y = "Monocytes", color = "Timepoint", palette = c("#00AFBB", "#E7B800", "#FC4E07"), FILL = "white", ylab = "Monocyte population", xlab = "Timepoint", add = "jitter", shape = "Timepoint"
)
bxp

stat.test <- cibersort2 %>%
  pairwise_wilcox_test(Monocytes ~ Timepoint) %>%
  add_significance()
stat.test

stat.test <- stat.test %>% add_xy_position(x = "Timepoint")
bxp +
  stat_pvalue_manual(stat.test, tip.length = 0) +
  labs(subtitle = get_test_label(stat.test, detailed = TRUE))

```

RNA-Seq Analysis

Import RNA-seq data
```{r}
Unnormalised <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\Unnormalised_text.txt")

PRENormalised <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\Normalised_text.txt")

```

```{r}
#Unnormalised (all patients -> 469D = patient 1)
Unnormalised2 <- Unnormalised[,-1]
rownames(Unnormalised2) <- Unnormalised[,1]
View(Unnormalised2)

#Normalised
PRENormalised2 <- PRENormalised[,-1]
rownames(PRENormalised2) <- PRENormalised[,1]
View(PRENormalised2)
```

Section 1 - Description of raw data 
-Basic bar graph showing count reads 
```{r}
#remove 10^e units
options(scipen = 999)
#plot the bar graph for number of counts per sample
barplot(colSums(Unnormalised2[,1:27]), main='Total Count Read Per Sample', xlab='Sample', ylab='Total Count', cex.names = 0.6, las = 2, cex.axis = 0.5)
#make total count in millions
barplot(colSums(Unnormalised2[,1:27])/1e06, names=colnames(Unnormalised2), las=2, ann=FALSE, cex.names=0.75)
mtext(side = 1, text = "Sample", line = 4)
mtext(side = 2, text = "Total Count (millions)", line = 3)
title("Barplot of Total Count Read Per Sample")

df1 <-colSums(Unnormalised2[,1:27])
is.data.frame(df1)
df1 <-as.data.frame(df1)
is.data.frame(df1)
df1$Samples <- rownames(df1)

Totalreads <-ggplot(data= df1, aes(x= Samples, y= df1 ,fill= Samples,))+
  geom_bar(stat="identity")+theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())

Totalreads

```

Section 1 - Description of raw data 
-Percentage of null counts per sample - expect to be similar within conditions 
```{r}
library(dplyr)
#Work out percentage of null counts in each column
PercentageOfNullCounts <- colSums(Unnormalised2==0)/nrow(Unnormalised2)*100
#Make a data frame so you can plot later on
PercentageOfNullCounts<- as.data.frame(PercentageOfNullCounts)
#Make another column for rownames
PercentageOfNullCounts$Sample <- rownames(PercentageOfNullCounts)

#Plot in bargraph
library(ggplot2)
PercentageOfNullCounts <-ggplot(PercentageOfNullCounts, aes(x=PercentageOfNullCounts, y=Sample)) +
  geom_bar(stat="identity")

PercentageOfNullCounts

```

Section 1 - Description of raw data 
-Log of this
-Density of counts distribution - again , expect density to be similar within conditions
```{r}
## The affy library has a density plotting function
library(affy)

## Create a list of 27 colours to use which are the same used throughout 
library(scales)
myColors <- hue_pal()(27)

## Plot the log2-transformed data with a 0.1 pseudocount
plotDensity(log2(Unnormalised2 + 0.1), col=rep(myColors, each=27),
            lty=c(1:ncol(Unnormalised2)), xlab='Log2(count)',
            main='Log density of counts distribution')

## Plot the raw data with a 0.1 pseudocount
plotDensity(log2(Unnormalised2), col=rep(myColors, each=27),xlab='Count',main='Density of counts distribution')
```

Section 1 - Description of raw data 
-Check to see if any of the reads is responsible for most of the reads 
-Percentage of reads associated with the sequences having the highest counts 
```{r}
Unnormalised2<-as.matrix(Unnormalised2)
View(Unnormalised2)
max <-colMaxs(Unnormalised2)
View(max)
colSums(Unnormalised2)

MaxReadCondition<- (colMaxs(Unnormalised2)/colSums(Unnormalised2))*100
View(MaxReadCondition)

MaxReadCondition<- as.data.frame(MaxReadCondition)
View(MaxReadCondition)

#make new condition as rownames
MaxReadCondition$Condition <- rownames(MaxReadCondition)
View(MaxReadCondition)

#group colours
group.colors <- c(X469D1="#333BFF",X469D2="#333BFF",X469D3="#333BFF",X479D1="#333BFF",X4792= "#333BFF",X479D3="#333BFF",X511D1="#333BFF", X511D2="#333BFF", X511D3="#333BFF",X469P11="#CC6600",X469P12="#CC6600",X469P13="#CC6600",X479P11="#CC6600",X479P12="#CC6600",X479P13="#CC6600",X511P14="#CC6600",X511P15="#CC6600",X511P16="#CC6600",X469P21="#9633FF", X469P22="#9633FF",X469P23="#9633FF",X479P21="#9633FF",X479P22="#9633FF",X479P23="#9633FF",X511P21="#9633FF", X511P22="#9633FF", X511P23="#9633FF")

#Plot in bargraph
MaxReadCondition<-ggplot(MaxReadCondition, aes(x=MaxReadCondition, y=Condition, fill = group.colors)) +
  geom_bar(stat="identity") + labs(fill = "Condition")

MaxReadCondition

```

PCA plot
```{r}
library(EDASeq)

#Plot the PCA 
EDASeq::plotPCA(Unnormalised2, cex = 0.6)

```

PCA plot after cpm normalisation 
```{r}
#Normalise the data and then plot the PCA 
cpmRaw <-cpm(Unnormalised2)

#Plot the PCA 
EDASeq::plotPCA(cpmRaw, cex = 0.6)
```

Section 2 - Variablilty of the experiment
-VST transformation
-Cluster dendrogram & PCA 
```{r}
#Transformation of the count data by Variance Stabilizing Transformation (VST) <- removes batch effect
library(DESeq2)
VSTcondition <- DESeq2::varianceStabilizingTransformation(Unnormalised2, blind = TRUE, fitType = "parametric")
#Plot the cluster dendrogram 
VSTcondition_transpose = t(VSTcondition)
#Euclidean distance
dist <- dist(VSTcondition_transpose, diag=TRUE)
#Hierarchical Clustering with hclust
hc <- hclust(dist)
#Plot the result
plot(hc)

#Plot PCA
EDASeq::plotPCA(VSTcondition, cex = 0.6)
```

```{r}
plotPCA(VSTcondition, intgroup = c("patient", "group"))
pcaData <- plotPCA(VSTcondition, intgroup = c( "patient", "group"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))

```

Section 2 - Variablilty of the experiment
-rlog-transformation data
(rlog is less sensitive to size factors, which can be an issue when size factors vary widely)
-Cluster dendrogram
```{r}
#Transformation of the count data by r log
rLogCondition <- rlog(Unnormalised2, blind = TRUE, fitType = "parametric")

#Plot the cluster dendrogram 
rLogcondition_transpose = t(rLogCondition)
#Euclidean distance
rlogdist <- dist(rLogcondition_transpose, diag=TRUE)
#Hierarchical Clustering with hclust
rloghc <- hclust(rlogdist)
#Plot the result
plot(rloghc)
#^same as above for VST transformation

```

Section 3 - Normalisation
Normalization was performed with locfunc="median"
-Size factors vs total of reads graphs 
-Normalisation box plots - counts distribution 
```{r}
#normalisation using locfunc way and plotting
NormalisationFactors <- estimateSizeFactorsForMatrix(
  Unnormalised2,
  locfunc = stats::median,
  type = c("ratio", "poscounts"))
NormalisationFactors <- as.data.frame(NormalisationFactors)
NormalisationTranspose <- t(NormalisationFactors)

#Normalized read counts are obtained by dividing raw read counts by the scaling factor associated with the sample they belong to. 
NormalisationOFUnnormalised2 <- rbind(Unnormalised2,NormalisationTranspose)
NormLocFunc <- data.matrix(NormalisationOFUnnormalised2) / rep(unlist(NormalisationOFUnnormalised2[39399,]), each=nrow(NormalisationOFUnnormalised2))

#Plot a PCA and clustering - to compare normalised to unnormalised (which was without transforming the data)

#PCA with normalisation
EDASeq::plotPCA(NormLocFunc, cex = 0.5)

#Transpose data
#Plot clustering (hclust)
NormLocFunc_transpose = t(NormLocFunc)
#Euclidean distance
ED <- dist(NormLocFunc_transpose, diag=TRUE)
#Hierarchical Clustering with hclust
HC <- hclust(ED)
#Plot the result
plot(HC)

#Plot box plots  
meltNormLocFunc<- melt(NormLocFunc)
meltNormLocFunc<- setNames(meltNormLocFunc, c("Read","Condition","Count"))
ggplot(meltNormLocFunc, aes(Condition, log(Count))) +
    geom_boxplot() +
    scale_y_continuous("Readout")+theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())

#Plot Box plots of raw data  
meltnorm<- melt(Unnormalised2)
meltnorm<- setNames(meltnorm, c("Read","Condition","Count"))
ggplot(meltnorm, aes(Condition, log(Count))) +
    geom_boxplot() +
    scale_y_continuous("Readout")+theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())

#Plot box plots of pre-normalised data - should look the same as plot 1. 
#rename columns 
colnames(PRENormalised2) <- c("X469D1","X469D2","X469D3","X479D1","X479D2","X479D3","X511D1", "X511D2", "X511D3", "X469P11","X469P12","X469P13", "X479P11", "X479P12", "X479P13","X511P14", "X511P15","X511P16","X469P21","X469P22", "X469P23", "X479P21", "X479P22","X479P23","X511P21", "X511P22","X511P23")

meltpre<- melt(PRENormalised2)
meltpre<- setNames(meltpre, c("Condition","Readnumber"))

ggplot(meltpre, aes(Condition, log(Readnumber))) +
    geom_boxplot() +
    scale_y_continuous("Readout")+theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())

```

Section 3 - Normalisation 
-size factor VS total count
```{r}
#Size factor vs total number of reads 
library(janitor)
Unnormalised3 <- Unnormalised %>% adorn_totals("row")
#Then set rows and columns 
Unnormalised4 <- Unnormalised3[,-1]
rownames(Unnormalised4) <- Unnormalised3[,1]

#Select just the rows you want 
normalisation_transpose = t(NormalisationFactors)
Unnormalised5 <- Unnormalised4[c(39399),]
row.names(Unnormalised5)[1] <- "total"  

#Make a new data frame 
RA = t(Unnormalised5)
SizeVsCount <- merge(RA, NormalisationFactors, by=0, all=TRUE)

library(ggplot2)
#Scatter plot
ggplot(SizeVsCount, aes(x=NormalisationFactors, y=total/10E5)) +
  geom_point(size=2, shape=17)  + geom_smooth(method = "lm", se = FALSE) +
        labs(title = "Size factor vs total number of reads",
             x = "Size Factors",
             y = "Total Number of Reads (millions)") + geom_text(aes(label=Row.names), size=3)  

#Afonso - used log
#replace y=total with log(total)
```

Section 3 - Normalisation 
-CPM normalisation 
```{r}
library(edgeR)
library(limma)
NEW<-cpm(Unnormalised2)

#Plot a PCA and clustering without 
#transpose data 
Condition_transpose2 = t(NEW)
#Euclidean distance
dist2 <- dist(Condition_transpose2, diag=TRUE)
#Hierarchical Clustering with hclust
hc2 <- hclust(dist2)
#Plot the result
plot(hc2)
```

Section 3 - Normalisation 
Data they provide that was normalised
```{r}
#Pre-normalised data 
#transpose data 
Condition_transpose3 = t(PRENormalised2)
#Euclidean distance
dist3 <- dist(Condition_transpose3, diag=TRUE)
#Hierarchical Clustering with hclust
hc3 <- hclust(dist3)
#Plot the result
plot(hc3)
```

Section 4 Differential analysis 
DESEQ 
```{r}
library(DESeq2)
#Filter the count matrix
Unnormalised2
View(Unnormalised2)

setwd(dir)

meta1<-read.table("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\sartools_metadata_all2.txt",row.names = 1,header=TRUE,stringsAsFactors = TRUE)
meta1

#Create the deseq object, taking into account the group (timepoint) and patient
DGEdeseq2_object <- DESeqDataSetFromMatrix(Unnormalised2, colData = meta1, design = ~Group+Patient)

#Perform Deseq
DGEdeseq2_object <- DESeq(DGEdeseq2_object)

#Convert to a data frame and then tidy up the results 
DGEdeseq2_results <- results(DGEdeseq2_object, contrast = c("Group","T2","T1"))
DGEdeseq2_results <- lfcShrink(DGEdeseq2_object, contrast = c("Group","T2","T1"), res=DGEdeseq2_results, type = 'normal')

#Remove the na values
DGEdeseq2_results <- na.omit(DGEdeseq2_results) 
DGEdeseq2_results <- DGEdeseq2_results[order(DGEdeseq2_results$padj),]
DGEdeseq2_results2 <- as.data.frame(DGEdeseq2_results)

#save results in a table
DGEdeseq2_results2 <- data.frame(DGEdeseq2_results2)
write.table(DGEdeseq2_results2, file="\\Users\\Maham\\Documents\\Project 1\\DGEdeseq2_results2.2.txt", sep = ",")

plot(DGEdeseq2_results$log2FoldChange,DGEdeseq2_results$padj)

```


Section 4 Differential analysis 
DESEQ - raw p values graphs  
```{r}
hist(DGEdeseq2_results$pvalue[DGEdeseq2_results$baseMean > 1], main="Histogram of P-values T2 vs T1", xlab="P-value", ylab="Frequency",breaks = 0:20/20,
     col = "green", border = "white")
```

Section 4 Differential analysis 
Results
MA plot (shows results for all contrasts)
```{r}
#Produce an MA plot 
#NB here need to call MA plot from the DESEQ package 
DESeq2::plotMA(DGEdeseq2_object)

#save
png(file = "MAPlot1.png")
DESeq2::plotMA(DGEdeseq2_object)
dev.off()

#Another MA plot for shrunken log fold changes - which remove the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds
DESeq2::plotMA(DGEdeseq2_object, ylim=c(-2,2))

#save
png(file = "MAPlot2.png")
DESeq2::plotMA(DGEdeseq2_object, ylim=c(-2,2))
dev.off()

DESeq2::plotMA(DGEdeseq2_object, ylim=c(-7,7))

#save
png(file = "MAPlot3.png")
DESeq2::plotMA(DGEdeseq2_object, ylim=c(-7,7))
dev.off()

```

Section 4 Differential analysis 
Results
Volcano plot
```{r, fig.width = 6, fig.height=8}
  EnhancedVolcano(DGEdeseq2_results,
    lab = rownames(DGEdeseq2_results),
    x = 'log2FoldChange',
    y = 'padj',
    ylab = '-Log10(padj)',
    title = 'T2 vs T1',
    pCutoff = 0.05,
    FCcutoff = FALSE,
    pointSize = 2.0,
    labSize = 4.0,
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)

#save volcano plot
  png(file = "VolcanoPlot.png")
  EnhancedVolcano(DGEdeseq2_results,
    lab = rownames(DGEdeseq2_results),
    x = 'log2FoldChange',
    y = 'padj',
    ylab = '-Log10(padj)',
    title = 'T2 vs T1',
    pCutoff = 0.05,
    FCcutoff = FALSE,
    pointSize = 2.0,
    labSize = 4.0,
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 1)
  dev.off()
```

Section 4 Differential analysis 
Results
-Filter genes to only select for those that are statistically significant and log2 fold change greater than 2
-Order the genes into top upregulated 
```{r}
#Filter out any log2 fold changes that are less than 2 & p-values that are greater than 0.5
#log2FC increase more than 2
DGEdeseq2_resultsFilteredINCREASE <- DGEdeseq2_results[DGEdeseq2_results$log2FoldChange > 2,]
#only statistically significant
DGEdeseq2_resultsFilteredINCREASE <- DGEdeseq2_resultsFilteredINCREASE[DGEdeseq2_resultsFilteredINCREASE$pvalue < 0.05,]
#log2FC decrease more than 2 & statistically significant
DGEdeseq2_resultsFilteredDECREASE<- DGEdeseq2_results[DGEdeseq2_results$log2FoldChange < -2,]
DGEdeseq2_resultsFilteredDECREASE <- DGEdeseq2_resultsFilteredDECREASE[DGEdeseq2_resultsFilteredDECREASE$pvalue < 0.05,]

library(dplyr)
library(tidyverse)
library(caret)
DGEdeseq2_resultsFilteredINCREASEupdated <- DGEdeseq2_resultsFilteredINCREASE
DGEdeseq2_resultsFilteredINCREASEupdated2 <- as.data.frame(DGEdeseq2_resultsFilteredINCREASEupdated)

DGEdeseq2_resultsFilteredINCREASEupdated2

#Ranking the genes in order of significance
#Create new column that multiplies the (1 - p-value) by log2 fold change
DGEdeseq2_resultsFilteredINCREASEupdated2["OneMinusPvalue"] <- (1 - DGEdeseq2_resultsFilteredINCREASEupdated2["pvalue"]) * DGEdeseq2_resultsFilteredINCREASEupdated2["log2FoldChange"]

#Rank the gene reads 
DGEdeseq2_resultsFilteredINCREASEupdated2<- DGEdeseq2_resultsFilteredINCREASEupdated2[order(DGEdeseq2_resultsFilteredINCREASEupdated2$OneMinusPvalue, decreasing=TRUE, na.last=TRUE),]

DGEdeseq2_resultsFilteredINCREASEupdated2

#save the table
OneMinusPvalueResultsTable <- data.frame(DGEdeseq2_resultsFilteredINCREASEupdated2)
write.table(OneMinusPvalueResultsTable, file="\\Users\\Maham\\Documents\\Project 1\\OneMinusPvalueResultsTable.txt", sep = ",")

```

Section 4 Differential analysis 
Top 20 most varied reads - bar graph 
```{r}
#Select the top 20 most varied genes
DGEtop20Increase2 <- DGEdeseq2_resultsFilteredINCREASEupdated2[1:20,]

#Make a simple bar graph 
#firstly, make the gene read column an actual column 
DGEtop20Increase2 <- cbind(Read = rownames(DGEtop20Increase2), DGEtop20Increase2)
#plot the bar graph 
p <-ggplot(data=DGEtop20Increase2, aes(x=Read, y=OneMinusPvalue)) +
  geom_bar(stat="identity")
p + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#save the bar plot
png(file = "BargraphTop20Reads.png")
p <-ggplot(data=DGEtop20Increase2, aes(x=Read, y=OneMinusPvalue)) +
  geom_bar(stat="identity")
p + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
dev.off()
```

Heatmap of DESEQ data 
```{r}
library(ggplot2)
library(scales) # needed for oob parameter
library(viridis)

# Transform count data using the variance stabilizing transform
DGEdeseq2_objectVST <- vst(DGEdeseq2_object)

# Convert the DESeq transformed object to a data frame
DGEdeseq2_objectVST <- assay(DGEdeseq2_objectVST)
DGEdeseq2_objectVST <- as.data.frame(DGEdeseq2_objectVST)
DGEdeseq2_objectVST$Gene <- rownames(DGEdeseq2_objectVST)
head(DGEdeseq2_objectVST)

# Keep only the significantly differentiated genes where the log2 fold change was greater than 3
sigGenes <- rownames(DGEdeseq2_results[DGEdeseq2_results$padj <= .05 & abs(DGEdeseq2_results$log2FoldChange) > 3,])
DGEdeseq2_objectVST <- DGEdeseq2_objectVST[DGEdeseq2_objectVST$Gene %in% sigGenes,]

# Convert the VST counts to long format for ggplot2
library(reshape2)

# First compare wide vs long version
DGEdeseq2_objectVST_wide <- DGEdeseq2_objectVST
DGEdeseq2_objectVST_long <- melt(DGEdeseq2_objectVST, id.vars=c("Gene"))

head(DGEdeseq2_objectVST_wide)
head(DGEdeseq2_objectVST_long)

# Now overwrite our original data frame with the long format
DGEdeseq2_objectVST <- melt(DGEdeseq2_objectVST, id.vars=c("Gene"))

library(ggplot2)

# Make a heatmap
heatmap <- ggplot(DGEdeseq2_objectVST, aes(x=variable, y=Gene, fill=value)) + geom_raster() + scale_fill_viridis(trans="sqrt") + theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())
heatmap

#save the heatmap
png(file = "Heatmap.png")
heatmap <- ggplot(DGEdeseq2_objectVST, aes(x=variable, y=Gene, fill=value)) + geom_raster() + scale_fill_viridis(trans="sqrt") + theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())
heatmap
dev.off()
```

Heatmap - add cluster dendrogram to it
```{r}
# Convert the significant genes back to a matrix for clustering
DGEdeseq2_objectVSTMatrix <- dcast(DGEdeseq2_objectVST, Gene ~ variable)
rownames(DGEdeseq2_objectVSTMatrix) <- DGEdeseq2_objectVSTMatrix$Gene
DGEdeseq2_objectVSTMatrix$Gene <- NULL

# Compute a distance calculation on both dimensions of the matrix
distanceGene <- dist(DGEdeseq2_objectVSTMatrix)
distanceSample <- dist(t(DGEdeseq2_objectVSTMatrix))

# Cluster based on the distance calculations
clusterGene <- hclust(distanceGene, method="average")
clusterSample <- hclust(distanceSample, method="average")

# Construct a dendrogram for samples
library(ggdendro)
sampleModel <- as.dendrogram(clusterSample)
sampleDendrogramData <- segment(dendro_data(sampleModel, type = "rectangle"))
sampleDendrogram <- ggplot(sampleDendrogramData) + geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) + theme_dendro()

# Re-factor samples for ggplot2
DGEdeseq2_objectVST$variable <- factor(DGEdeseq2_objectVST$variable, levels=clusterSample$labels[clusterSample$order])

# Construct the heatmap - note that at this point we have only clustered the samples NOT the genes
heatmap <- ggplot(DGEdeseq2_objectVST, aes(x=variable, y=Gene, fill=value)) + geom_raster() + scale_fill_viridis(trans="sqrt") + theme(axis.text.x=element_text(angle=65, hjust=1), axis.text.y=element_blank(), axis.ticks.y=element_blank())
heatmap

# Combine the dendrogram and the heatmap
library(gridExtra)
grid.arrange(sampleDendrogram, heatmap, ncol=1, heights=c(1,5))
```

Line up the dendrogram
```{r}
# Load in libraries necessary for modifying plots
library(gtable)
library(grid)
library(ggplot2)

# Modify the ggplot objects
sampleDendrogram_1 <- sampleDendrogram + scale_x_continuous(expand=c(.0085, .0085)) + scale_y_continuous(expand=c(0, 0))
heatmap_1 <- heatmap + scale_x_discrete(expand=c(0, 0)) + scale_y_discrete(expand=c(0, 0))

# Convert both grid based objects to grobs
sampleDendrogramGrob <- ggplotGrob(sampleDendrogram_1)
heatmapGrob <- ggplotGrob(heatmap_1)

# Check the widths of each grob
sampleDendrogramGrob$widths
heatmapGrob$widths

# Add in the missing columns
sampleDendrogramGrob <- gtable_add_cols(sampleDendrogramGrob, heatmapGrob$widths[7], 6)
sampleDendrogramGrob <- gtable_add_cols(sampleDendrogramGrob, heatmapGrob$widths[8], 7)

# Make sure every width between the two grobs is the same
maxWidth <- unit.pmax(sampleDendrogramGrob$widths, heatmapGrob$widths)
sampleDendrogramGrob$widths <- as.list(maxWidth)
heatmapGrob$widths <- as.list(maxWidth)

# Arrange the grobs into a plot
finalGrob <- arrangeGrob(sampleDendrogramGrob, heatmapGrob, ncol=1, heights=c(2,5))

# Draw the plot
plot(finalGrob)

# Save the plot
png(file = "HeatmapClustered.png")
plot(finalGrob)
dev.off()

```

M2 macrophage signature genes
```{r, fig.width = 12, fig.height = 5}

#Import LM22 file which only contains M2 macrophage signature genes (already sorted from high to low expression)
M2_LM22 <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\M2.txt")
M2_LM22 <- data.frame(M2_LM22)
head(M2_LM22)

#select just gene names column from LM22 
M2_LM22_Degs <- select(M2_LM22, c(1))
head(M2_LM22_Degs)

DESeq_patients_Degs <- row.names(DGEdeseq2_results2)
DESeq_patients_Degs <- data.frame(DESeq_patients_Degs)
colnames(DESeq_patients_Degs)[1] <- "Gene.symbol"

#Find which M2 macrophage signature genes in LM22 are also present in the results from differential gene expression between T2 and T1
Matched_genes <- merge(DESeq_patients_Degs, M2_LM22_Degs, sort = FALSE)

Matched_genes <- data.frame(Matched_genes)
write.table(Matched_genes, file="\\Users\\Maham\\Documents\\Project 1\\Matched_genes.txt", sep = ",")

#Top 30 expressed M2 macrophage signature genes from LM22 in T2vsT1 DGE
Top30_sig_expressed_genes_LM22andT2vsT1 <- DGEdeseq2_results2[c(M2_LM22_Degs[c(1:30),]),]

#save as table
Top30_sig_expressed_genes_LM22andT2vsT1 <- data.frame(Top30_sig_expressed_genes_LM22andT2vsT1)
write.table(Top30_sig_expressed_genes_LM22andT2vsT1, file="\\Users\\Maham\\Documents\\Project 1\\Top30_sig_expressed_genes_LM22andT2vsT1.txt", sep = ",")


##Below is extra

#Top50 in LM22
Top50_sig_expressed_genes_LM22andT2vsT1 <- DGEdeseq2_results2[c(M2_LM22_Degs[c(1:50),]),]

#save as table
Top50_sig_expressed_genes_LM22andT2vsT1 <- data.frame(Top50_sig_expressed_genes_LM22andT2vsT1)
write.table(Top50_sig_expressed_genes_LM22andT2vsT1, file="\\Users\\Maham\\Documents\\Project 1\\Top50_sig_expressed_genes_LM22andT2vsT1.txt", sep = ",")

#Top 30 matched genes (in CIBERSORTx signature matrix (LM22) and T2vsT1 DESEQ results)
Top30_Matched_genes <- DGEdeseq2_results2[c("GZMH", "GZMA", "GZMK", "LY86", "S1PR5", "CCL5", "IL7R", "GPR183", "ASGR2", "GZMB", "PLA2G7", "MS4A1", "DUSP2", "CD1D", "CD19", "CD2", "CD86", "NLRP3", "MYB", "UPK3A", "CR2", "PIK3IP1", "TXK", "IL2RB", "IRF8", "CD244", "CD1C", "ABCB4", "CD4", "IFNG"),]
#LogFC > 2
Top30_Matched_genesFilteredINCREASE <- Top30_Matched_genes[Top30_Matched_genes$log2FoldChange > 2,]
#only statistically significant
Top30_Matched_genesFilteredINCREASE <- Top30_Matched_genesFilteredINCREASE[Top30_Matched_genesFilteredINCREASE$pvalue < 0.05,]
#save as table
Top30_Matched_genesFilteredINCREASE_Table <- data.frame(Top30_Matched_genesFilteredINCREASE)
write.table(Top30_Matched_genesFilteredINCREASE_Table, file="\\Users\\Maham\\Documents\\Project 1\\Top30_Matched_genesFilteredINCREASE_Table.txt", sep = ",")

#check ALL matched genes in DESEQ output for T2vsT1, LogFC > 2
All_Matched_genes <- DGEdeseq2_results2[c(Matched_genes[c(1:469),]),]
All_Matched_genesFilteredINCREASE <- All_Matched_genes[All_Matched_genes$log2FoldChange > 2,]
#only statistically significant
All_Matched_genesFilteredINCREASE <- All_Matched_genesFilteredINCREASE[All_Matched_genesFilteredINCREASE$pvalue < 0.05,]
#save as table
All_Matched_genesFilteredINCREASE_Table <- data.frame(All_Matched_genesFilteredINCREASE)
write.table(All_Matched_genesFilteredINCREASE_Table, file="\\Users\\Maham\\Documents\\Project 1\\All_Matched_genesFilteredINCREASE_Table.txt", sep = ",")

#test genes with Log2FC > 1
All_Matched_genes2 <- DGEdeseq2_results2[c(Matched_genes[c(1:469),]),]
All_Matched_genesFilteredINCREASE2 <- All_Matched_genes2[All_Matched_genes2$log2FoldChange > 1,]
#only statistically significant
All_Matched_genesFilteredINCREASE2 <- All_Matched_genesFilteredINCREASE2[All_Matched_genesFilteredINCREASE2$pvalue < 0.05,]
#save as table
All_Matched_genesFilteredINCREASE2_Table <- data.frame(All_Matched_genesFilteredINCREASE2)
write.table(All_Matched_genesFilteredINCREASE2_Table, file="\\Users\\Maham\\Documents\\Project 1\\All_Matched_genesFilteredINCREASE2_Table.txt", sep = ",")

#check if the matched genes are macrophage markers
All_LM22 <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\LM22_Matched_genes.txt")
All_LM22 <- data.frame(All_LM22)
head(All_LM22)

All_LM22New <- All_LM22[,-1]
rownames(All_LM22New) <- All_LM22[,1]
All_LM22New <- data.frame(All_LM22New)

#LogFC > 2
ALL_Matched_genesFilteredINCREASE_LM22 <- All_LM22New[c(rownames(All_Matched_genesFilteredINCREASE)),] 
ALL_Matched_genesFilteredINCREASE_LM22 <- as.matrix(ALL_Matched_genesFilteredINCREASE_LM22)
#save as table
ALL_Matched_genesFilteredINCREASE_LM22_Table <- data.frame(ALL_Matched_genesFilteredINCREASE_LM22)
write.table(ALL_Matched_genesFilteredINCREASE_LM22_Table, file="\\Users\\Maham\\Documents\\Project 1\\ALL_Matched_genesFilteredINCREASE_LM22_Table.txt", sep = ",")

#LogFC > 1
ALL_Matched_genesFilteredINCREASE_LM22_2 <- All_LM22New[c(rownames(All_Matched_genesFilteredINCREASE2)),] 
ALL_Matched_genesFilteredINCREASE_LM22_2 <- as.matrix(ALL_Matched_genesFilteredINCREASE_LM22_2)
#save as table
ALL_Matched_genesFilteredINCREASE_LM22_2_Table <- data.frame(ALL_Matched_genesFilteredINCREASE_LM22_2)
write.table(ALL_Matched_genesFilteredINCREASE_LM22_2_Table, file="\\Users\\Maham\\Documents\\Project 1\\ALL_Matched_genesFilteredINCREASE_LM22_2_Table.txt", sep = ",")

#Just focusing on Monocytes/Macrophages
M_LM22New <- select(All_LM22New, c(13:16))

#LogFC > 2
M_Matched_genesFilteredINCREASE_LM22 <- M_LM22New[c(rownames(All_Matched_genesFilteredINCREASE)),] 
M_Matched_genesFilteredINCREASE_LM22 <- as.matrix(M_Matched_genesFilteredINCREASE_LM22)
#save as table
M_Matched_genesFilteredINCREASE_LM22_Table <- data.frame(M_Matched_genesFilteredINCREASE_LM22)
write.table(M_Matched_genesFilteredINCREASE_LM22_Table, file="\\Users\\Maham\\Documents\\Project 1\\M_Matched_genesFilteredINCREASE_LM22_Table.txt", sep = ",")

#LogFC > 1
M_Matched_genesFilteredINCREASE_LM22_2 <- M_LM22New[c(rownames(All_Matched_genesFilteredINCREASE2)),] 
M_Matched_genesFilteredINCREASE_LM22_2 <- as.matrix(M_Matched_genesFilteredINCREASE_LM22_2)
#save as table
M_Matched_genesFilteredINCREASE_LM22_2_Table <- data.frame(M_Matched_genesFilteredINCREASE_LM22_2)
write.table(M_Matched_genesFilteredINCREASE_LM22_2_Table, file="\\Users\\Maham\\Documents\\Project 1\\M_Matched_genesFilteredINCREASE_LM22_2_Table.txt", sep = ",")

#Top30 most significant genes present in LM22 and DESeq T2vsT1
Top30_sig_Matched_genes <- read.delim("\\Users\\Maham\\Desktop\\Aliza\\Imperial\\Project\\Bespoke_project\\Data\\RNASeq\\Top30_sig_Matched_genes.txt")
Top30_sig_Matched_genesNew <- Top30_sig_Matched_genes[,-1]
rownames(Top30_sig_Matched_genesNew) <- Top30_sig_Matched_genes[,1]
Top30_sig_Matched_genesNew <- data.frame(Top30_sig_Matched_genesNew)

Top30_sig_Matched_genesNew_LM22 <- All_LM22New[c(rownames(Top30_sig_Matched_genesNew)),] 
Top30_sig_Matched_genesNew_LM22 <- as.matrix(Top30_sig_Matched_genesNew_LM22)
#save as table
Top30_sig_Matched_genesNew_LM22_Table <- data.frame(Top30_sig_Matched_genesNew_LM22)
write.table(Top30_sig_Matched_genesNew_LM22_Table, file="\\Users\\Maham\\Documents\\Project 1\\Top30_sig_Matched_genesNew_LM22_Table.txt", sep = ",")

```

Plot bar graph of top 30 M2 macrophage signature genes in T2 vs T1
```{r, fig.width = 12, fig.height = 5}

#Important Bar graph 
Top30_sig_expressed_genes_LM22andT2vsT1$Gene <- rownames(Top30_sig_expressed_genes_LM22andT2vsT1)
Top30_sig_expressed_genes_LM22andT2vsT1New <- na.omit(Top30_sig_expressed_genes_LM22andT2vsT1)
plot <- ggplot(Top30_sig_expressed_genes_LM22andT2vsT1New, aes(fill=padj<0.05, x=Gene, y=log2FoldChange))+
geom_bar(position="dodge", stat="identity", width=0.5) + ggtitle("M2 Macrophage Genes Expressed in T2 vs T1") +
  xlab("Gene") +
  ylab("Log2FoldChange")
plot + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1), axis.text = element_text(size = 15), axis.title = element_text(size = 20), plot.title = element_text(size = 20), legend.text = element_text(size = 10), legend.title = element_text(size = 15))

```

** Exploratory analysis and visualisation (another way) **
```{r}
#vst and rlog 
#Sequencing depth correction is done automatically for vst and rlog
DGEdeseq2_objectNEW <- DESeqDataSetFromMatrix(Unnormalised2, colData = meta1, design = ~ Patient + Group)

vsd <- vst(DGEdeseq2_objectNEW, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)

rld <- rlog(DGEdeseq2_objectNEW, blind = FALSE)
head(assay(rld), 3)

library("dplyr")
library("ggplot2")

dds <- estimateSizeFactors(DGEdeseq2_objectNEW)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)

#sample distances
sampleDists <- dist(t(assay(vsd)))
sampleDists

library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$Group, vsd$Patient, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

#using Poisson Distance
library("PoiClaClu")
poisd <- PoissonDistance(t(counts(dds)))
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste(dds$Group, dds$Patient, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)

#PCA plot using VST data
plotPCA(vsd, intgroup = c("Group", "Patient"))
pcaData <- plotPCA(vsd, intgroup = c("Group", "Patient"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))
PCAplot <- ggplot(pcaData, aes(x = PC1, y = PC2, color = Group, shape = Patient)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")
PCAplot + theme(axis.text.x = element_text( vjust = 1, hjust=1), axis.text = element_text(size = 15), axis.title = element_text(size = 15), plot.title = element_text(size = 20), legend.text = element_text(size = 10), legend.title = element_text(size = 15))

#MDS plot using VST data
mds <- as.data.frame(colData(vsd))  %>%
         cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = Group, shape = Patient)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with VST data")

#MDS with Poisson Distances
mdsPois <- as.data.frame(colData(dds)) %>%
   cbind(cmdscale(samplePoisDistMatrix))
ggplot(mdsPois, aes(x = `1`, y = `2`, color = Group, shape = Patient)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with Poisson Distances")
```

DGE (another way)
```{r}
dds <- DESeq(dds)
res <- results(dds, contrast=c("Group","T2","T1"))
res
mcols(res, use.names = TRUE)
summary(res)

#padj threshold
res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)

#lfc threshold
resLFC1 <- results(dds, lfcThreshold=2)
table(resLFC1$padj < 0.05)

sum(res$padj < 0.05, na.rm=TRUE)
resSig <- subset(res, padj < 0.05)
#subset results table to these genes ^ and then sort by log2foldchange - strongest down-regulation
head(resSig[order(resSig$log2FoldChange), ])
# - strongest up-regulation
head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])
```

Plotting Results (another way)
```{r}
#Looking at the expression levels of genes of interest across timepoints
topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup=c("Group"))

library("ggbeeswarm")
geneCounts <- plotCounts(dds, gene = "CCL18", intgroup = c("Group","Patient"),
                         returnData = TRUE)
ggplot(geneCounts, aes(x = Group, y = count, color = Patient)) +
  scale_y_log10() +  geom_beeswarm(cex = 3)
#^get top 10 genes (from M2 signature genes bar graph)

#Normalized counts with lines connecting different patient samples
GenePlot <- ggplot(geneCounts, aes(x = Group, y = count, color = Patient, group = Patient)) +
  scale_y_log10() + geom_point(size = 3) + geom_line() + ggtitle("CCL18") +
  xlab("Group") +
  ylab("Count")
GenePlot + theme(axis.text.x = element_text( vjust = 1, hjust=1), axis.text = element_text(size = 15), axis.title = element_text(size = 20), plot.title = element_text(size = 20), legend.text = element_text(size = 10), legend.title = element_text(size = 15))

#MA-plot (or mean-difference plot)
library("ashr")
resultsNames(dds)

#shrinks noisy log2 fold changes
resAsh <- lfcShrink(dds, contrast=c("Group", "T2", "T1"), type="ashr")
plotMA(resAsh, ylim = c(-5, 5))

#label individual points on MA-plot
plotMA(resAsh, ylim = c(-5,5))
topGene <- rownames(resAsh)[which.min(res$padj)]
with(resAsh[topGene, ], {
  points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
  text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
})

#Histogram of p-values
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")

#Gene Clustering
library("genefilter")
#clustering only relevant for genes which actually carry a signal; usually only cluster a subset of most highly variable genes
#select top 20 genes with highest variance across samples (VST data)
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)

#Heatmap of relative VST-transformed values across samples
mat  <- assay(vsd)[topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd)[, c("Patient","Group")])
pheatmap(mat, annotation_col = anno)

```

GSEA
```{r}
library(fgsea)
library(httr)

#Remove the non significant values 
DGEdeseq2_resultsFilteredINCREASEupdated2

#Make a new data frame to play with
gseaInput <- DGEdeseq2_resultsFilteredINCREASEupdated2

#Need to make another column ranking the data
#use rank by logFC
gseaInput$Symbol <- rownames(gseaInput)
gseaInput$Rank <- c(1:length(gseaInput$Symbol))

gseaInputRanked <- gseaInput[,c("Rank", "Symbol")]
gseaInputRanked <- as.data.frame(gseaInputRanked)

#Read in pathway set
PathwaySet <- gmtPathways("\\Users\\Maham\\Documents\\Project 1\\h.all.v7.4.symbols.gmt")

#Run Gene set enrichment 
ranks1 <- setNames(gseaInputRanked$Rank,gseaInputRanked$Symbol)

gsea <- fgsea(PathwaySet,ranks1)
gsea

```

GSEA by
(1 - p-value) x log2 fold change
```{r}
#New data frame to play with
Samples_1pvalueXlfc <- DGEdeseq2_results
Samples_1pvalueXlfc <- as.data.frame(Samples_1pvalueXlfc)

#Create new column that multiplies the (1 - p-value) by log2 fold change 
Samples_1pvalueXlfc["OneMinusPvalue_LFC"] <- (1 - Samples_1pvalueXlfc["pvalue"]) * Samples_1pvalueXlfc["log2FoldChange"]

#Make the gene names a column and rank a column
Samples_1pvalueXlfc$symbol <- rownames(Samples_1pvalueXlfc)

#Just select the gene and rank columns 
Samples_1pvalueXlfc_GSEAinput <- Samples_1pvalueXlfc[,c("OneMinusPvalue_LFC","symbol")]

#set seed 
set.seed(5656)

#Read in pathway set
PathwaySet <- gmtPathways("\\Users\\Maham\\Documents\\Project 1\\h.all.v7.4.symbols.gmt")
names(PathwaySet) <- gsub("_", " ", names(PathwaySet))
names(PathwaySet) <- stringr::str_to_title(names(PathwaySet), locale = "en")

#Run Gene set enrichment 
Samples_1pvalueXlfc_GSEAinputNEW <- setNames(Samples_1pvalueXlfc_GSEAinput$OneMinusPvalue_LFC,Samples_1pvalueXlfc_GSEAinput$symbol)
gsea_1pvalueXlfc <- fgsea(PathwaySet,Samples_1pvalueXlfc_GSEAinputNEW) 

#Select top 30 genes 
Samples_1pvalueXlfc_GSEAinputNEW_top30 <- gsea_1pvalueXlfc[order(gsea_1pvalueXlfc$NES ,decreasing=TRUE,na.last=TRUE),]
Samples_1pvalueXlfc_GSEAinputNEW_top30 <- Samples_1pvalueXlfc_GSEAinputNEW_top30[1:30,]

#Plot 
ggplot(Samples_1pvalueXlfc_GSEAinputNEW_top30, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

#Now try it ranked 
Samples_1pvalueXlfc_GSEAinput_ranked <- Samples_1pvalueXlfc_GSEAinput[order(Samples_1pvalueXlfc_GSEAinput$OneMinusPvalue_LFC ,decreasing=TRUE,na.last=TRUE),]

#set seed 
set.seed(5656)

#Run Gene set enrichment 
Samples_1pvalueXlfc_GSEAinput_ranked_new <- setNames(Samples_1pvalueXlfc_GSEAinput_ranked$OneMinusPvalue_LFC,Samples_1pvalueXlfc_GSEAinput_ranked$symbol)
gsea_1pvalueXlfc_ranked <- fgsea(PathwaySet,Samples_1pvalueXlfc_GSEAinput_ranked_new) 

#Select top 30 genes 
Samples_1pvalueXlfc_GSEAinputNEW_top30_ranked <- gsea_1pvalueXlfc_ranked[order(gsea_1pvalueXlfc_ranked$NES ,decreasing=TRUE,na.last=TRUE),]
Samples_1pvalueXlfc_GSEAinputNEW_top30_ranked <- Samples_1pvalueXlfc_GSEAinputNEW_top30_ranked[1:30,]

#Plot 
ggplot(Samples_1pvalueXlfc_GSEAinputNEW_top30_ranked, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

```

GSEA by
p-value x sign of log2 fold change   
```{r}
#New data frame to play with
Samples_deseq_pvalue_and_sign <- DGEdeseq2_results
Samples_deseq_pvalue_and_sign <- as.data.frame(Samples_deseq_pvalue_and_sign)

#Create a new column that has the p-value multiplied by the sign of the log2 fold change 
Samples_deseq_pvalue_and_sign$sign_log2FoldChange <- sign(Samples_deseq_pvalue_and_sign$log2FoldChange)
Samples_deseq_pvalue_and_sign["SignLogXPvalue"] <- Samples_deseq_pvalue_and_sign["pvalue"] * Samples_deseq_pvalue_and_sign["sign_log2FoldChange"]

#Make the gene names a column and rank a column
Samples_deseq_pvalue_and_sign$symbol <- rownames(Samples_deseq_pvalue_and_sign)

#Just select the gene and rank columns 
Samples_deseq_pvalue_and_sign_GSEAinput <- Samples_deseq_pvalue_and_sign[,c("SignLogXPvalue","symbol")]

#set seed
set.seed(5656)

#Run Gene set enrichment 
Samples_deseq_pvalue_and_sign_GSEAinputNEW <- setNames(Samples_deseq_pvalue_and_sign_GSEAinput$SignLogXPvalue,Samples_deseq_pvalue_and_sign_GSEAinput$symbol)
gsea_pvalueXsign <- fgsea(PathwaySet,Samples_deseq_pvalue_and_sign_GSEAinputNEW)

#Select top 30 genes 
gsea_pvalueXsign_top30 <- gsea_pvalueXsign[order(gsea_pvalueXsign$NES ,decreasing=TRUE,na.last=TRUE),]
gsea_pvalueXsign_top30 <- gsea_pvalueXsign_top30[1:30,]

#Plot 
ggplot(gsea_pvalueXsign_top30, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

```

** GSEA using stat and no ordering **
-this one used
```{r}
STAT_samples <- DGEdeseq2_results
STAT_samples$symbol <- rownames(STAT_samples)
STAT_samples_new <- STAT_samples[,c("stat","symbol")]

#Read in pathway set
PathwaySet <- gmtPathways("\\Users\\Maham\\Documents\\Project 1\\h.all.v7.4.symbols.gmt")
names(PathwaySet) <- gsub("_", " ", names(PathwaySet))
names(PathwaySet) <- stringr::str_to_title(names(PathwaySet), locale = "en")

#set seed
set.seed(5656)

#Run Gene set enrichment 
ranks_STAT_new <- setNames(STAT_samples_new$stat,STAT_samples_new$symbol)
gsea_stat_new <- fgsea(PathwaySet,ranks_STAT_new)

#Just get significant pathways 
gsea_stat_new_significant <- gsea_stat_new[gsea_stat_new$padj < 0.05,]
gsea_stat_new_significant

#rank and select top 30 upregulated gene sets 
gsea_stat_new_significant_top30<- gsea_stat_new_significant[order(gsea_stat_new_significant$NES ,decreasing=TRUE,na.last=TRUE),]
#Bar graph of top 20 most varied genes
gsea_stat_new_significant_top30 <- gsea_stat_new_significant_top30[1:30,]

#Plot 
ggplot(gsea_stat_new_significant_top30, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark Pathways NES from GSEA T2 vs T1") + 
  theme_minimal()

```

Shrink the LFC and then perform GSEA 
```{r}

resultsNames(DGEdeseq2_object)
res <- lfcShrink(DGEdeseq2_object, coef=2)
plotMA(res)
res2 <- as.data.frame(res)
#Remove the na values
res2 <- na.omit(res2) 

#make rownames a column
res2$symbol <- rownames(res2)

STAT_samples <- DGEdeseq2_results
STAT_samples$symbol <- rownames(STAT_samples)

#select columns to perform the GSEA on
res2_gsea_input <- STAT_samples[,c("log2FoldChange","symbol")]

#Read in pathway set
PathwaySet <- gmtPathways("\\Users\\Maham\\Documents\\Project 1\\h.all.v7.4.symbols.gmt")
names(PathwaySet) <- gsub("_", " ", names(PathwaySet))
names(PathwaySet) <- stringr::str_to_title(names(PathwaySet), locale = "en")

#set seed
set.seed(5656)

#Run Gene set enrichment 
ranks_LFC_shrink <- setNames(res2_gsea_input$log2FoldChange,res2_gsea_input$symbol)
gsea_LFC_shrink<- fgsea(PathwaySet,ranks_LFC_shrink)

#Just get significant pathways 
gsea_LFC_shrink_signif <- gsea_LFC_shrink[gsea_LFC_shrink$padj < 0.05,]
gsea_LFC_shrink_signif

#rank and select top 30 upregulated gene sets 
gsea_LFC_shrink_signif_top30<- gsea_LFC_shrink_signif[order(gsea_LFC_shrink_signif$NES ,decreasing=TRUE,na.last=TRUE),]
#Bar graph of top 20 most varied genes
gsea_LFC_shrink_signif_top30 <- gsea_LFC_shrink_signif_top30[1:30,]

#Plot 
ggplot(gsea_LFC_shrink_signif_top30, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=pval<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

```

GSEA by
log2(adjusted p-value) * sign of the log2 fold change
```{r}
#New data frame to play with
Samples_deseq_log2AdjPvalue_signlfc <- DGEdeseq2_results
Samples_deseq_log2AdjPvalue_signlfc <- as.data.frame(Samples_deseq_log2AdjPvalue_signlfc)
Samples_deseq_log2AdjPvalue_signlfc

#Create a new column that has the log2(adj p-value) multiplied by the sign of the log2 fold change
Samples_deseq_log2AdjPvalue_signlfc$sign_log2FoldChange <- sign(Samples_deseq_log2AdjPvalue_signlfc$log2FoldChange)
Samples_deseq_log2AdjPvalue_signlfc$log2AdjPvalue <- log2(Samples_deseq_log2AdjPvalue_signlfc$padj)
Samples_deseq_log2AdjPvalue_signlfc["log2padjXSignlfc"] <- Samples_deseq_log2AdjPvalue_signlfc["log2AdjPvalue"] * Samples_deseq_log2AdjPvalue_signlfc["sign_log2FoldChange"]


#Make the gene names a column and rank a column
Samples_deseq_log2AdjPvalue_signlfc$symbol <- rownames(Samples_deseq_log2AdjPvalue_signlfc)

#Just select the gene and rank columns 
Samples_deseq_log2AdjPvalue_signlfc_GSEAinput <- Samples_deseq_log2AdjPvalue_signlfc[,c("log2padjXSignlfc","symbol")]
Samples_deseq_log2AdjPvalue_signlfc_GSEAinput2 <- Samples_deseq_log2AdjPvalue_signlfc_GSEAinput

#get rid of genes with na values
Samples_deseq_log2AdjPvalue_signlfc_GSEAinputNEW <- na.omit(Samples_deseq_log2AdjPvalue_signlfc_GSEAinput2)

#set seed
set.seed(5656)

#Read in pathway set
PathwaySet <- gmtPathways("\\Users\\Maham\\Documents\\Project 1\\All_T2vsT1_gene_sets.gmt")
names(PathwaySet) <- gsub("_", " ", names(PathwaySet))
names(PathwaySet) <- stringr::str_to_title(names(PathwaySet), locale = "en")

#Run Gene set enrichment 
Samples_deseq_log2AdjPvalue_signlfc_GSEAinputNEW <- setNames(Samples_deseq_log2AdjPvalue_signlfc_GSEAinput$log2padjXSignlfc,Samples_deseq_log2AdjPvalue_signlfc_GSEAinput$symbol)
gsea_log2padjXSignlfc<- fgsea(PathwaySet,Samples_deseq_log2AdjPvalue_signlfc_GSEAinputNEW)

#Select top 30 genes 
gsea_log2padjXSignlfc_top <- gsea_log2padjXSignlfc[order(gsea_log2padjXSignlfc$NES,decreasing=TRUE,na.last=TRUE),]
gsea_log2padjXSignlfc_top30 <- gsea_log2padjXSignlfc_top[1:30,]

#Plot 
ggplot(gsea_log2padjXSignlfc_top30, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Pathways NES from GSEA") + 
  theme_minimal()

```

